components/cqrs/event_processor_group.go

```diff
@@ -0,0 +1,260 @@
+		err = stdErrors.Join(err, errors.New("missing Marshaler"))
+	}
+
+	if c.GenerateSubscribeTopic == nil {
+		err = stdErrors.Join(err, errors.New("missing GenerateHandlerGroupTopic"))
+	}
+	if c.SubscriberConstructor == nil {
+		err = stdErrors.Join(err, errors.New("missing SubscriberConstructor"))
+	}
+
+	return err
+}
+
+type EventGroupProcessorGenerateSubscribeTopicFn func(EventGroupProcessorGenerateSubscribeTopicParams) (string, error)
+
+type EventGroupProcessorGenerateSubscribeTopicParams struct {
+	EventGroupName     string
+	EventGroupHandlers []GroupEventHandler
+}
+
+type EventGroupProcessorSubscriberConstructorFn func(EventGroupProcessorSubscriberConstructorParams) (message.Subscriber, error)
+
+type EventGroupProcessorSubscriberConstructorParams struct {
+	EventGroupName     string
+	EventGroupHandlers []GroupEventHandler
+}
+
+type EventGroupProcessorOnHandleFn func(params EventGroupProcessorOnHandleParams) error
+
+type EventGroupProcessorOnHandleParams struct {
+	GroupName string
+	Handler   GroupEventHandler
+
+	Event     any
+	EventName string
+
+	// Message is never nil and can be modified.
+	Message *message.Message
+}
+
+// EventGroupProcessor determines which EventHandler should handle event received from event bus.
+// Compared to EventProcessor, EventGroupProcessor allows to have multiple handlers that share the same subscriber instance.
+type EventGroupProcessor struct {
+	groupEventHandlers map[string][]GroupEventHandler
+
+	config EventGroupProcessorConfig
+}
+
+// NewEventGroupProcessorWithConfig creates a new EventGroupProcessor.
+func NewEventGroupProcessorWithConfig(config EventGroupProcessorConfig) (*EventGroupProcessor, error) {
+	config.setDefaults()
+
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config EventProcessor")
+	}
+
+	return &EventGroupProcessor{
+		groupEventHandlers: map[string][]GroupEventHandler{},
+		config:             config,
+	}, nil
+}
+
+// AddHandlersGroup adds a new list of GroupEventHandler to the EventProcessor.
+//
+// Compared to AddHandlers, AddHandlersGroup allows to have multiple handlers that share the same subscriber instance.
+//
+// IMPORTANT: It's required to call AddHandlersToRouter to add the handlers to the router after calling AddHandlersGroup.
+//
+// Handlers group needs to be unique within the EventProcessor instance.
+func (p *EventGroupProcessor) AddHandlersGroup(handlerName string, handlers ...GroupEventHandler) error {
+	if len(handlers) == 0 {
+		return errors.New("no handlers provided")
+	}
+	if _, ok := p.groupEventHandlers[handlerName]; ok {
+		return fmt.Errorf("event handler group '%s' already exists", handlerName)
+	}
+
+	p.groupEventHandlers[handlerName] = handlers
+
+	return nil
+}
+
+// AddHandlersToRouter adds the EventProcessor's handlers to the given router.
+// It should be called only once per EventProcessor instance.
+func (p EventGroupProcessor) AddHandlersToRouter(r *message.Router) error {
```

Comment:
How about passing a router instance to `NewEventGroupProcessorWithConfig` and having `AddHandlersGroup` call `AddHandlersToRouter` internally? It could be a simpler design to avoid having to remember to call the method.

Reply:
good idea! done here: 7fe0ca7